cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -1057,7 +1057,11 @@ macro(build_boost)
   set(BOOST_VENDORED TRUE)
 endmacro()
 
-if(ARROW_BUILD_TESTS)
+if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION
+                                              VERSION_GREATER_EQUAL 16)
+  # GH-34094 we require newer Boost versions for Clang 16 or newer
```

Comment:
I'd add another comment line about how older boost versions use the long-deprecated std::unary_function in boost/container_hash/hash.hpp and that support has been removed from clang 16. That way we know why this is here and under what conditions we can remove it.

Reply:
AH yes that's a good clarification, thanks